Sources/ServiceDiscovery/InMemoryServiceDiscovery.swift

```diff
@@ -12,209 +12,91 @@
-            let subscriptions = self._serviceSubscriptions
-            self._serviceSubscriptions = [:]
-            self._serviceInstances = [:]
-            return subscriptions.values
-        }
+        private var underlying: AsyncThrowingStream<[Instance], Error>
 
-        guard let serviceSubscriptions = maybeServiceSubscriptions else {
-            return
+        init(_ underlying: AsyncThrowingStream<[Instance], Error>) {
+            self.underlying = underlying
         }
 
-        for subscriptions in serviceSubscriptions {
-            for sub in subscriptions.lazy.filter({ !$0.cancellationToken.isCancelled }) {
-                sub.completionHandler(.serviceDiscoveryUnavailable)
-            }
+        public func makeAsyncIterator() -> AsyncIterator {
+            AsyncIterator(self.underlying.makeAsyncIterator())
         }
-    }
 
-    private func _lookupNow(_ service: Service) -> [Instance]? {
-        if let instances = self._serviceInstances[service] {
-            return instances
-        } else {
-            return nil
-        }
-    }
+        public struct AsyncIterator: AsyncIteratorProtocol {
+            private var underlying: AsyncThrowingStream<[Instance], Error>.Iterator
 
-    private struct Subscription {
-        let nextResultHandler: (Result<[Instance], Error>) -> Void
-        let completionHandler: (CompletionReason) -> Void
-        let cancellationToken: CancellationToken
-    }
-}
-
-public extension InMemoryServiceDiscovery {
-    struct Configuration {
-        /// Default configuration
-        public static var `default`: Configuration {
-            .init()
-        }
-
-        /// Lookup timeout in case `deadline` is not specified
-        public var defaultLookupTimeout: DispatchTimeInterval = .milliseconds(100)
-
-        internal var serviceInstances: [Service: [Instance]]
-
-        public init() {
-            self.init(serviceInstances: [:])
-        }
-
-        /// Initializes `InMemoryServiceDiscovery` with the given service to instances mappings.
-        public init(serviceInstances: [Service: [Instance]]) {
-            self.serviceInstances = serviceInstances
-        }
+            init(_ underlying: AsyncThrowingStream<[Instance], Error>.Iterator) {
+                self.underlying = underlying
+            }
 
-        /// Registers `service` and its `instances`.
-        public mutating func register(service: Service, instances: [Instance]) {
-            self.serviceInstances[service] = instances
+            public mutating func next() async throws -> [Instance]? {
+                try await self.underlying.next()
+            }
         }
     }
 }
 
-// MARK: - NSLock extensions
-
-private extension NSLock {
-    func withLock<Result>(_ body: () throws -> Result) rethrows -> Result {
-        self.lock()
-        defer {
-            self.unlock()
-        }
-        return try body()
+#if swift(<5.9)
+// Async stream API backfil
+public extension AsyncThrowingStream {
```

Comment:
this is an oversight. This should not be public. Also swiftformat seems to be configured wrong here.
```suggestion
extension AsyncThrowingStream {
```

Reply:
this still needs fixing.